### PR TITLE
[token-2022] Update confidential transfer `Transfer` and `TransferWithFee` for Solana 1.16

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
Confidential transfer `Transfer` and `TransferWithFee` instructions are not updated for the new Solana 1.16 changes.

#### Summary of changes
- [bdaf1fe](https://github.com/solana-labs/solana-program-library/pull/4779/commits/bdaf1fe9baf70944146f556bc3c98663898ee97b): The logic for the account info types needed for instructions was getting long, so I refactored these from `state.rs` into a separate module

- [e88c9e4](https://github.com/solana-labs/solana-program-library/pull/4779/commits/e88c9e4031e1bd912c3c934a7231f7ae322e0b26): I updated the processor and client logic for the `Transfer` instruction. This should be pretty standard except that I added helper functions that extract the relevant bits from the transfer amount and fee ciphertexts. The transfer amount and fee are encrypted under the source, destination, auditor, and/or withdraw withheld authority ElGamal pubkeys simultaneously. These helper functions extract the relevant components that are relevant for the source, destination, and/or withdraw withheld authority. The helper functions belong more naturally in the zk-token-sdk, but I inlined these for now until these are added on the monorepo side.

- [b07d480](https://github.com/solana-labs/solana-program-library/pull/4779/commits/b07d48076c3126317bbec5896b767ba8fdad1a04): The same set of changes as the previous commit for the `TransferWithFee` instruction.

- [2ce0429](https://github.com/solana-labs/solana-program-library/pull/4779/commits/2ce0429c0c4d63f3c35c500494461022e8902013): Pretty straightforward updates to the `Transfer` instruction tests.

- [8150248](https://github.com/solana-labs/solana-program-library/pull/4779/commits/8150248724406c0323d78039f0609961f98c3a65): The same set of changes as the previous commit for the `Transfer` with memo tests.

For this PR, I did not update the tests for the `TransferWithFee` since that would involve initializing the mint and account for the confidential transfer fee extension, which probably belong in a separate PR.

### CODE REVIEW

File: token/program-2022/src/extension/confidential_transfer/account_info.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
+#[cfg(not(target_os = "solana"))]
```
Comment (joncinque) [inline token/program-2022/src/extension/confidential_transfer/account_info.rs:1]: nit for simplicity: does it make sense to mark this whole file with `#![cfg(not(target_os = "solana")))]`? and not have all of the individual annotations? or even mark the module declaration in `confidential_transfer/mod.rs` with the cfg?

File: token/program-2022/src/extension/confidential_transfer/processor.rs [hunk 2]
```diff
@@ -586,70 +554,123 @@ fn process_transfer(
...
+/// Extract the transfer amount ciphertext encrypted under the source ElGamal public key.
+#[cfg(feature = "zk-ops")]
+fn transfer_amount_source_ciphertext(
+    transfer_amount_ciphertext: &TransferAmountCiphertext,
+) -> ElGamalCiphertext {
+    let transfer_amount_ciphertext_bytes = bytemuck::bytes_of(transfer_amount_ciphertext);
+
+    let mut source_ciphertext_bytes = [0u8; 64];
+    source_ciphertext_bytes[..32].copy_from_slice(&transfer_amount_ciphertext_bytes[..32]);
+    source_ciphertext_bytes[32..].copy_from_slice(&transfer_amount_ciphertext_bytes[32..64]);
```
Comment (joncinque) [inline token/program-2022/src/extension/confidential_transfer/processor.rs:617]: nit: This is definitely done purposefully  -- can you add comments about specifically which bits you're extracting for these lines?

File: token/program-2022/src/extension/confidential_transfer/processor.rs [hunk 3]
```diff
@@ -586,70 +554,123 @@ fn process_transfer(
...
+/// Extract the transfer amount ciphertext encrypted under the destination ElGamal public key.
+#[cfg(feature = "zk-ops")]
+fn transfer_amount_destination_ciphertext(
+    transfer_amount_ciphertext: &TransferAmountCiphertext,
+) -> ElGamalCiphertext {
+    let transfer_amount_ciphertext_bytes = bytemuck::bytes_of(transfer_amount_ciphertext);
+
+    let mut destination_ciphertext_bytes = [0u8; 64];
+    destination_ciphertext_bytes[..32].copy_from_slice(&transfer_amount_ciphertext_bytes[..32]);
+    destination_ciphertext_bytes[32..].copy_from_slice(&transfer_amount_ciphertext_bytes[64..96]);
```
Comment (joncinque) [inline token/program-2022/src/extension/confidential_transfer/processor.rs:631]: Same here, just a little comment about which parts you're copying

File: token/program-2022-test/tests/confidential_transfer.rs [hunk 4]
```diff
@@ -1106,7 +1103,7 @@ async fn confidential_transfer_withdraw() {
         .await;
 
     // attempt to withdraw without enough funds
-    let err = token
+    token
```
Comment (joncinque) [inline token/program-2022-test/tests/confidential_transfer.rs:1106]: Can you check that the error is correct here?

File: token/program-2022-test/tests/confidential_transfer.rs [hunk 5]
```diff
@@ -1237,43 +1229,16 @@ async fn ct_transfer() {
...
-        err,
-        TokenClientError::Client(Box::new(TransportError::TransactionError(
-            TransactionError::InstructionError(
-                0,
-                InstructionError::Custom(
-                    TokenError::MaximumPendingBalanceCreditCounterExceeded as u32
-                ),
-            )
-        )))
-    );
```
Comment (joncinque) [inline token/program-2022-test/tests/confidential_transfer.rs:1272]: This seems like an important case to check, can you add it back in?

**Review by joncinque — APPROVED** [review]
Looks really great! Really just a couple of nits to cleanup, but nothing that should stop this from going in after that

File: token/program-2022/src/extension/confidential_transfer/account_info.rs (hunk 1 above)
Comment (samkim-crypto) [reply token/program-2022/src/extension/confidential_transfer/account_info.rs:1]: Ah yes, this will be much simpler! I marked the module declaration to be `#![cfg(not(target_os = "solana")))]`.

File: token/program-2022/src/extension/confidential_transfer/processor.rs (hunk 2 above)
Comment (samkim-crypto) [reply token/program-2022/src/extension/confidential_transfer/processor.rs:617]: Done!

File: token/program-2022-test/tests/confidential_transfer.rs (hunk 5 above)
Comment (samkim-crypto) [reply token/program-2022-test/tests/confidential_transfer.rs:1272]: This check is verifying that a confidential transfer fails if the maximum credit limit in the destination account is reached. In practice the default destination will be `2^16`, but in order to test this, we need to set the destination account maximum credit limit to be low (say `2`). 

In order to simplify the tests, to initialize the `alice_meta`, I used the helper constructor  `ConfidentialTokenAccountMeta::new_with_tokens`, which deposits some funds to `alice`'s account. The `new_with_tokens` interface unfortunately does not allow us to set custom maximum credit limit. Instead of adding a new constructor `ConfidentialTokenAccountMeta::new_with_tokens_and_maximum_pending_balance_credit_counter`, I just initialize `bob_meta` with the constructor `ConfidentialTokenAccountMeta::new_with_maximum_pending_balance_credit_counter` and tested that a transfer to `bob`'s account fails if the maximum credit counter is reached.

So in summary, this check on `alice`'s account is not removed, but it is just moved to work on `bob`'s account purely instead for simplicity.

File: token/program-2022-test/tests/confidential_transfer.rs (hunk 4 above)
Comment (samkim-crypto) [reply token/program-2022-test/tests/confidential_transfer.rs]: Yes, now I remember why I didn't check for the error. I forgot to mention it in the PR description :pensive: . For some reason, the error assert does not hold even though both the left and the right hand side of the assert is `TokenClientError::ProofGeneration`. No matter what I try, I can't seem to make the assert to be true.

For testing, I tried changing the returned [error type](https://github.com/solana-labs/solana-program-library/pull/4779/files#diff-465a63e60c1758459d46845310375141037eda9b9043efe9a44f3e55063ac72eR1981) to be any other variant to `TokenClientError` (and correspondingly in the assert statement) and the assert statement holds true. It seems like there is a very weird behavior going on with `TokenClientError::ProofGeneration`. Have you encountered this type of behavior anywhere else?

Comment (joncinque) [reply token/program-2022-test/tests/confidential_transfer.rs]: oh, you know what, this might be because of the error equality in `TokenError`, which does a goofy string equality. This is entirely my fault. If you want to implement something better, please do! Either way, your change is good.

Comment (joncinque) [reply token/program-2022-test/tests/confidential_transfer.rs]: https://github.com/solana-labs/solana-program-library/blob/4e13218b41211a9a83397b4f6dea0135241a102f/token/client/src/token.rs#L92-L116 just add the missing ones!

**Review by joncinque — APPROVED** [review]
You'll probably just need to add those cases for the TokenClientError equality check. Once that's in and this passes CI, this is good to go on my side!

File: token/program-2022-test/tests/confidential_transfer.rs (hunk 4 above)
Comment (samkim-crypto) [reply token/program-2022-test/tests/confidential_transfer.rs]: Oh, this was right below where the error type is define and I missed it! Thanks for the pointer! I think the CI should pass now.
